platform/posix/transport/src/openssl_posix.c

```diff
@@ -0,0 +1,563 @@
+        {
+            LogDebug( ( "Attempting to open client's certificate: Path=%.*s.",
+                        ( int32_t ) pOpensslCredentials->rootCaPathLen,
+                        pOpensslCredentials->pRootCaPath ) );
+        }
+        else
+        {
+            LogDebug( ( "Attempting to open client's certificate: Path=%s/%.*s.",
+                        cwd,
+                        ( int32_t ) pOpensslCredentials->rootCaPathLen,
+                        pOpensslCredentials->pRootCaPath ) );
+        }
+
+        sslStatus = SSL_CTX_use_certificate_chain_file( pSslContext,
+                                                        pOpensslCredentials->pClientCertPath );
+
+        /* Import the client certificate. */
+        if( sslStatus != 1 )
+        {
+            LogError( ( "SSL_CTX_use_certificate_chain_file failed to import "
+                        "client certificate at %.*s.",
+                        ( int32_t ) pOpensslCredentials->clientCertPathLen,
+                        pOpensslCredentials->pClientCertPath ) );
+        }
+        else
+        {
+            LogDebug( ( "Successfully imported client certificate." ) );
+        }
+    }
+
+    if( ( sslStatus == 1 ) &&
+        ( pOpensslCredentials->pPrivateKeyPath != NULL ) &&
+        ( pOpensslCredentials->privateKeyPathLen > 0 ) )
+    {
+        /* Log the absolute directory based on first character of client certificate path. */
+        if( ( pOpensslCredentials->pPrivateKeyPath[ 0 ] == '/' ) || ( pOpensslCredentials->pPrivateKeyPath[ 0 ] == '\\' ) )
+        {
+            LogDebug( ( "Attempting to open client's private key: Path=%.*s.",
+                        ( int32_t ) pOpensslCredentials->privateKeyPathLen,
+                        pOpensslCredentials->pPrivateKeyPath ) );
+        }
+        else
+        {
+            LogDebug( ( "Attempting to open client's private key: Path=%s/%.*s.",
+                        cwd,
+                        ( int32_t ) pOpensslCredentials->privateKeyPathLen,
+                        pOpensslCredentials->pPrivateKeyPath ) );
+        }
+
+        sslStatus = SSL_CTX_use_PrivateKey_file( pSslContext,
+                                                 pOpensslCredentials->pPrivateKeyPath,
+                                                 SSL_FILETYPE_PEM );
+
+        /* Import the client certificate private key. */
+        if( sslStatus != 1 )
+        {
+            LogError( ( "SSL_CTX_use_PrivateKey_file failed to import client "
+                        "certificate private key at %.*s.",
+                        ( int32_t ) pOpensslCredentials->privateKeyPathLen,
+                        pOpensslCredentials->pPrivateKeyPath ) );
+        }
+        else
+        {
+            LogDebug( ( "Successfully imported client certificate private key." ) );
+        }
+    }
+
+    /* Free the root CA object. */
+    if( pRootCa != NULL )
+    {
+        X509_free( pRootCa );
```

Comment:
Are we sure we want to free `pRootCa` here?  It has been added to the ssl context cert store at line 143 to be used later.  According to [the documentation](https://www.openssl.org/docs/manmaster/man3/X509_STORE_add_cert.html), _"X509_STORE_add_cert() and X509_STORE_add_crl() add the respective object to the X509_STORE's local storage. Untrusted objects should not be added in this way. The added object's reference count is incremented by one, hence __the caller retains ownership of the object__ and needs to free it when it is no longer needed."_

Based on that description, it seems like we would want to delete these objects when the SSL context is destroyed since the root certificate would be needed for authentication.

Reply:
Good point. I have removed this. Based on my research, calling `SSL_CTX_free` should already handle for us when calling `Openssl_Disconnect`.